property setters/delers should have implicit `None` return type even when `default_return` is disabled

like with `__init__`

```py
class A:
    @property
    def foo(self) -> int:
        ...
    @foo.setter
    def foo(self, value: int): # error: Function is missing a return type annotation  [no-untyped-def]
        ...
```

Also the type of `value` should be inferrable from the getter.

--- a/.mypy/baseline.json
+++ b/.mypy/baseline.json
@@ -2783,7 +2783,7 @@
         "code": "no-any-expr",
         "column": 49,
         "message": "Expression type contains \"Any\" (has type \"type[CallableType]\")",
-        "offset": 7,
+        "offset": 8,
         "target": "mypy.checker.TypeChecker.check_func_def"
       },
       {
@@ -2795,9 +2795,23 @@
       },
       {
         "code": "no-any-expr",
+        "column": 68,
+        "message": "Expression type contains \"Any\" (has type \"type[CallableType]\")",
+        "offset": 5,
+        "target": "mypy.checker.TypeChecker.check_func_def"
+      },
+      {
+        "code": "no-any-expr",
+        "column": 68,
+        "message": "Expression type contains \"Any\" (has type \"type[CallableType]\")",
+        "offset": 0,
+        "target": "mypy.checker.TypeChecker.check_func_def"
+      },
+      {
+        "code": "no-any-expr",
         "column": 63,
         "message": "Expression type contains \"Any\" (has type \"type[CallableType]\")",
-        "offset": 51,
+        "offset": 59,
         "target": "mypy.checker.TypeChecker.check_func_def"
       },
       {
@@ -2825,7 +2839,7 @@
         "code": "no-any-expr",
         "column": 39,
         "message": "Expression type contains \"Any\" (has type \"type[CallableType]\")",
-        "offset": 16,
+        "offset": 17,
         "target": "mypy.checker.TypeChecker.check_for_missing_annotations"
       },
       {
@@ -3173,9 +3187,30 @@
       },
       {
         "code": "no-any-expr",
+        "column": 40,
+        "message": "Expression type contains \"Any\" (has type \"type[CallableType]\")",
+        "offset": 621,
+        "target": "mypy.checker.TypeChecker.visit_decorator"
+      },
+      {
+        "code": "no-any-expr",
+        "column": 85,
+        "message": "Expression type contains \"Any\" (has type \"type[CallableType]\")",
+        "offset": 4,
+        "target": "mypy.checker.TypeChecker.visit_decorator"
+      },
+      {
+        "code": "no-any-expr",
+        "column": 85,
+        "message": "Expression type contains \"Any\" (has type \"type[CallableType]\")",
+        "offset": 0,
+        "target": "mypy.checker.TypeChecker.visit_decorator"
+      },
+      {
+        "code": "no-any-expr",
         "column": 33,
         "message": "Expression type contains \"Any\" (has type \"type[CallableType]\")",
-        "offset": 1884,
+        "offset": 1271,
         "target": "mypy.checker.TypeChecker.check_subtype"
       },
       {
@@ -14390,9 +14425,37 @@
       },
       {
         "code": "no-any-expr",
+        "column": 58,
+        "message": "Expression type contains \"Any\" (has type \"type[CallableType]\")",
+        "offset": 112,
+        "target": "mypy.nodes.FuncItem.is_dynamic"
+      },
+      {
+        "code": "no-any-expr",
+        "column": 58,
+        "message": "Expression type contains \"Any\" (has type \"type[CallableType]\")",
+        "offset": 0,
+        "target": "mypy.nodes.FuncItem.is_dynamic"
+      },
+      {
+        "code": "no-any-expr",
+        "column": 58,
+        "message": "Expression type contains \"Any\" (has type \"type[CallableType]\")",
+        "offset": 0,
+        "target": "mypy.nodes.FuncItem.is_dynamic"
+      },
+      {
+        "code": "no-any-expr",
+        "column": 58,
+        "message": "Expression type contains \"Any\" (has type \"type[CallableType]\")",
+        "offset": 0,
+        "target": "mypy.nodes.FuncItem.is_dynamic"
+      },
+      {
+        "code": "no-any-expr",
         "column": 15,
         "message": "Expression type contains \"Any\" (has type \"dict[str, Any]\")",
-        "offset": 160,
+        "offset": 49,
         "target": "mypy.nodes.FuncDef.serialize"
       },
       {
@@ -16629,7 +16692,7 @@
         "code": "no-any-expr",
         "column": 41,
         "message": "Expression type contains \"Any\" (has type \"type[CallableType]\")",
-        "offset": 645,
+        "offset": 644,
         "target": "mypy.semanal.SemanticAnalyzer.analyze_func_def"
       },
       {
@@ -16718,9 +16781,30 @@
       },
       {
         "code": "no-any-expr",
+        "column": 64,
+        "message": "Expression type contains \"Any\" (has type \"type[CallableType]\")",
+        "offset": 138,
+        "target": "mypy.semanal.SemanticAnalyzer.analyze_property_with_multi_part_definition"
+      },
+      {
+        "code": "no-any-expr",
+        "column": 64,
+        "message": "Expression type contains \"Any\" (has type \"type[CallableType]\")",
+        "offset": 7,
+        "target": "mypy.semanal.SemanticAnalyzer.analyze_property_with_multi_part_definition"
+      },
+      {
+        "code": "no-any-expr",
+        "column": 64,
+        "message": "Expression type contains \"Any\" (has type \"type[CallableType]\")",
+        "offset": 17,
+        "target": "mypy.semanal.SemanticAnalyzer.analyze_property_with_multi_part_definition"
+      },
+      {
+        "code": "no-any-expr",
         "column": 31,
         "message": "Expression type contains \"Any\" (has type \"type[CallableType]\")",
-        "offset": 187,
+        "offset": 72,
         "target": "mypy.semanal.SemanticAnalyzer.check_classvar_in_signature"
       },
       {
@@ -17077,7 +17161,7 @@
         "code": "no-any-expr",
         "column": 37,
         "message": "Expression type contains \"Any\" (has type \"type[CallableType]\")",
-        "offset": 28,
+        "offset": 29,
         "target": "mypy.semanal.infer_fdef_types_from_defaults"
       },
       {

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -991,9 +991,18 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
                             arg_types.append(super_type.arg_types[super_i])
                     elif not defn.type:
                         arg_types.append(UntypedType())
+
                 if defn.type:
                     assert isinstance(defn.type, CallableType)
-                    if isinstance(get_proper_type(defn.type.ret_type), AnyType):
+                    if self.options.default_return and isinstance(defn.type.ret_type, NoneType):
+                        if defn.unanalyzed_type:
+                            assert isinstance(defn.unanalyzed_type, CallableType)
+                            if is_unannotated_any(get_proper_type(defn.unanalyzed_type.ret_type)):
+                                defn.type.ret_type = super_type.ret_type
+                        else:
+                            defn.type.ret_type = super_type.ret_type
+
+                    if is_unannotated_any(get_proper_type(defn.type.ret_type)):
                         defn.type.ret_type = super_type.ret_type
                 else:
                     typ = defn.type = CallableType(
@@ -4156,17 +4165,6 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
                     e.var.type = AnyType(TypeOfAny.special_form)
                     e.var.is_ready = True
                     return
-        # detect default_return and warn against it
-        #  This isn't perfect, as None return could come from inheritance, but who cares
-        if (
-            self.options.disallow_untyped_defs and e.var.is_property
-            and isinstance(e.func.type, CallableType)
-        ):
-            if isinstance(
-                get_proper_type(e.func.type.ret_type), NoneType
-            ) and not e.func.unanalyzed_type:
-                self.fail("Property is missing a type annotation",
-                          e.func, code=codes.NO_UNTYPED_DEF)
         if self.recurse_into_functions:
             with self.tscope.function_scope(e.func):
                 self.check_func_item(e.func, name=e.func.name)
@@ -4206,7 +4204,17 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
         if e.func.info and e.func.name in ('__init__', '__new__'):
             if e.type and not isinstance(get_proper_type(e.type), (FunctionLike, AnyType)):
                 self.fail(message_registry.BAD_CONSTRUCTOR_TYPE, e)
-
+        # detect default_return and warn against it
+        #  This isn't perfect, as None return could come from inheritance, but who cares
+        if (
+            self.options.disallow_untyped_defs and e.var.is_property
+            and isinstance(e.func.type, CallableType)
+        ):
+            if isinstance(
+                get_proper_type(e.func.type.ret_type), NoneType
+            ) and (not e.func.unanalyzed_type or is_unannotated_any(e.func.unanalyzed_type.ret_type)):
+                self.fail("Property is missing a type annotation",
+                          e.func, code=codes.NO_UNTYPED_DEF)
     def check_for_untyped_decorator(self,
                                     func: FuncDef,
                                     dec_type: Type,

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -1020,6 +1020,9 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
         #     typ = typ.copy_modified(arg_types = arg_types)
         #     defn.type = typ
 
+        if not defn.is_dynamic():
+            self.dynamic_funcs[-1] = False
+
         # Expand type variables with value restrictions to ordinary types.
         expanded = self.expand_typevars(defn, typ)
         for item, typ in expanded:
@@ -1259,7 +1262,8 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
         has_explicit_annotation = (isinstance(fdef.type, CallableType)
                                    and any(not is_unannotated_any(t)
                                            for t in fdef.type.arg_types + [fdef.type.ret_type]))
-
+        if fdef.type and fdef.is_dynamic():
+            return
         show_untyped = not self.is_typeshed_stub or self.options.warn_incomplete_stub
         check_incomplete_defs = self.options.disallow_incomplete_defs and has_explicit_annotation
         if show_untyped and (self.options.disallow_untyped_defs or check_incomplete_defs):
@@ -4152,7 +4156,17 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
                     e.var.type = AnyType(TypeOfAny.special_form)
                     e.var.is_ready = True
                     return
-
+        # detect default_return and warn against it
+        #  This isn't perfect, as None return could come from inheritance, but who cares
+        if (
+            self.options.disallow_untyped_defs and e.var.is_property
+            and isinstance(e.func.type, CallableType)
+        ):
+            if isinstance(
+                get_proper_type(e.func.type.ret_type), NoneType
+            ) and not e.func.unanalyzed_type:
+                self.fail("Property is missing a type annotation",
+                          e.func, code=codes.NO_UNTYPED_DEF)
         if self.recurse_into_functions:
             with self.tscope.function_scope(e.func):
                 self.check_func_item(e.func, name=e.func.name)

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -994,7 +994,9 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
 
                 if defn.type:
                     assert isinstance(defn.type, CallableType)
-                    if self.options.default_return and isinstance(defn.type.ret_type, NoneType):
+                    if self.options.default_return and isinstance(get_proper_type(
+                        defn.type.ret_type), NoneType
+                    ):
                         if defn.unanalyzed_type:
                             assert isinstance(defn.unanalyzed_type, CallableType)
                             if is_unannotated_any(get_proper_type(defn.unanalyzed_type.ret_type)):
@@ -4212,9 +4214,12 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
         ):
             if isinstance(
                 get_proper_type(e.func.type.ret_type), NoneType
-            ) and (not e.func.unanalyzed_type or is_unannotated_any(e.func.unanalyzed_type.ret_type)):
+            ) and (not e.func.unanalyzed_type or (isinstance(e.func.unanalyzed_type, CallableType)
+                and is_unannotated_any(e.func.unanalyzed_type.ret_type))
+            ):
                 self.fail("Property is missing a type annotation",
                           e.func, code=codes.NO_UNTYPED_DEF)
+
     def check_for_untyped_decorator(self,
                                     func: FuncDef,
                                     dec_type: Type,

--- a/mypy/nodes.py
+++ b/mypy/nodes.py
@@ -697,7 +697,8 @@ class FuncItem(FuncBase):
             arg.set_line(self.line, self.column, self.end_line)
 
     def is_dynamic(self) -> bool:
-        return self.type is None
+        from mypy.types import CallableType
+        return self.type is None or isinstance(self.type, CallableType) and self.type.implicit
 
 
 FUNCDEF_FLAGS: Final = FUNCITEM_FLAGS + [

--- a/mypy/semanal.py
+++ b/mypy/semanal.py
@@ -47,6 +47,7 @@ Some important properties:
   daemon strips and reuses existing ASTs (to improve performance and/or
   reduce memory use).
 """
+
 from contextlib import contextmanager
 
 from typing import (
@@ -989,6 +990,13 @@ class SemanticAnalyzer(NodeVisitor[None],
                 if len(item.decorators) == 1:
                     node = item.decorators[0]
                     if isinstance(node, MemberExpr):
+                        none_func = callable_type(
+                            item.func, self.named_type("builtins.function"), NoneType()
+                        )
+                        untyped_func = callable_type(
+                            item.func, self.named_type("builtins.function"),
+                            UntypedType(TypeOfAny.explicit)
+                        )
                         if node.name == 'setter':
                             # The first item represents the entire property.
                             first_item.var.is_settable_property = True
@@ -997,9 +1005,7 @@ class SemanticAnalyzer(NodeVisitor[None],
                             # infer types from the getter
                             if isinstance(first_item.func.type, CallableType):
                                 if not item.func.type:
-                                    item.func.type = callable_type(
-                                        item.func, self.named_type("builtins.function"), NoneType()
-                                    )
+                                    item.func.type = none_func
                                     if len(item.func.type.arg_types) == 2:
                                         item.func.type.arg_types[1] = first_item.func.type.ret_type
                                 elif isinstance(item.func.type, CallableType):
@@ -1013,10 +1019,7 @@ class SemanticAnalyzer(NodeVisitor[None],
                             else:
                                 # set some boring defaults
                                 if not item.func.type:
-                                    item.func.type = callable_type(
-                                        item.func, self.named_type("builtins.function"),
-                                        UntypedType(TypeOfAny.explicit)
-                                    )
+                                    item.func.type = untyped_func
                                     item.func.type.arg_types[1] = UntypedType(TypeOfAny.explicit)
                                     item.func.type.implicit = True
                                 elif isinstance(item.func.type, CallableType):
@@ -1031,15 +1034,10 @@ class SemanticAnalyzer(NodeVisitor[None],
                             # infer types from the getter
                             if first_item.func.type:
                                 if not item.type:
-                                    item.func.type = callable_type(
-                                        item.func, self.named_type("builtins.function"), NoneType()
-                                    )
+                                    item.func.type = none_func
                             else:
                                 if not item.func.type:
-                                    item.func.type = callable_type(
-                                        item.func, self.named_type("builtins.function"),
-                                        UntypedType(TypeOfAny.explicit)
-                                    )
+                                    item.func.type = untyped_func
                                     item.func.type.implicit = True
                 else:
                     self.fail("Decorated property not supported", item)

--- a/mypy/semanal.py
+++ b/mypy/semanal.py
@@ -47,7 +47,7 @@ Some important properties:
   daemon strips and reuses existing ASTs (to improve performance and/or
   reduce memory use).
 """
-
+from builtins import isinstance
 from contextlib import contextmanager
 
 from typing import (
@@ -101,7 +101,7 @@ from mypy.types import (
     PROTOCOL_NAMES, TYPE_ALIAS_NAMES, FINAL_TYPE_NAMES, FINAL_DECORATOR_NAMES, REVEAL_TYPE_NAMES,
     is_named_instance, is_unannotated_any, UntypedType,
 )
-from mypy.typeops import function_type, get_type_vars, infer_impl_from_parts
+from mypy.typeops import function_type, get_type_vars, infer_impl_from_parts, callable_type
 from mypy.type_visitor import TypeQuery
 from mypy.typeanal import (
     TypeAnalyser, analyze_type_alias, no_subscript_builtin_alias,
@@ -995,6 +995,53 @@ class SemanticAnalyzer(NodeVisitor[None],
                             first_item.var.is_settable_property = True
                             # Get abstractness from the original definition.
                             item.func.is_abstract = first_item.func.is_abstract
+                            # infer types from the getter
+                            if isinstance(first_item.func.type, CallableType):
+                                if not item.func.type:
+                                    item.func.type = callable_type(
+                                        item.func, self.named_type("builtins.function"), NoneType()
+                                    )
+                                    if len(item.func.type.arg_types) == 2:
+                                        item.func.type.arg_types[1] = first_item.func.type.ret_type
+                                elif isinstance(item.func.type, CallableType):
+                                    if (
+                                        len(item.func.type.arg_types) == 2
+                                        and is_unannotated_any(item.func.type.arg_types[1])
+                                    ):
+                                        item.func.type.arg_types[1] = first_item.func.type.ret_type
+                                    if is_unannotated_any(item.func.type.ret_type):
+                                        item.func.type.ret_type = NoneType()
+                            else:
+                                # set some boring defaults
+                                if not item.func.type:
+                                    item.func.type = callable_type(
+                                        item.func, self.named_type("builtins.function"),
+                                        UntypedType(TypeOfAny.explicit)
+                                    )
+                                    item.func.type.arg_types[1] = UntypedType(TypeOfAny.explicit)
+                                    item.func.type.implicit = True
+                                elif isinstance(item.func.type, CallableType):
+                                    if (
+                                        len(item.func.type.arg_types) == 2
+                                        and is_unannotated_any(item.func.type.arg_types[1])
+                                    ):
+                                        item.func.type.arg_types[1] = NoneType()
+                                    if is_unannotated_any(item.func.type.ret_type):
+                                        item.func.type.ret_type = NoneType()
+                        if node.name == 'deleter':
+                            # infer types from the getter
+                            if first_item.func.type:
+                                if not item.type:
+                                    item.func.type = callable_type(
+                                        item.func, self.named_type("builtins.function"), NoneType()
+                                    )
+                            else:
+                                if not item.func.type:
+                                    item.func.type = callable_type(
+                                        item.func, self.named_type("builtins.function"),
+                                        UntypedType(TypeOfAny.explicit)
+                                    )
+                                    item.func.type.implicit = True
                 else:
                     self.fail("Decorated property not supported", item)
                 item.func.accept(self)
@@ -5639,7 +5686,8 @@ def infer_fdef_types_from_defaults(defn: Union[FuncDef, Decorator], self: Semant
                                      ret_type or UntypedType(),
                                      self.named_type('builtins.function'),
                                      line=defn.line,
-                                     column=defn.column)
+                                     column=defn.column,
+                                     implicit=not self.options.disallow_untyped_defs)
     elif defn.type:
         assert isinstance(defn.type, CallableType)
         if self.options.default_return and is_unannotated_any(defn.type.ret_type) and (

--- a/mypy/semanal.py
+++ b/mypy/semanal.py
@@ -47,7 +47,6 @@ Some important properties:
   daemon strips and reuses existing ASTs (to improve performance and/or
   reduce memory use).
 """
-from builtins import isinstance
 from contextlib import contextmanager
 
 from typing import (

--- a/test-data/unit/check-based-infer-function-types.test
+++ b/test-data/unit/check-based-infer-function-types.test
@@ -339,7 +339,8 @@ del a.p
 
 
 [case testPropertyInheritInfer-xfail]
-# Overloaded overrides aren't supported yet
+# Overloaded overrides aren't supported yet, so this is only the getter
+# xfail because infer function types has been disabled for now.
 class A:
     @property
     def p(self) -> int: ...

--- a/test-data/unit/check-based-infer-function-types.test
+++ b/test-data/unit/check-based-infer-function-types.test
@@ -5,9 +5,11 @@
 
 def f(): ...
 reveal_type(f)  # N: Revealed type is "def () -> Untyped"
+f()  # E: Call to untyped function "f" in typed context  [no-untyped-call]
 
 def g(i=1, b=""): ...
 reveal_type(g)  # N: Revealed type is "def (i: int =, b: str =) -> Untyped"
+g()  # E: Call to untyped function "g" in typed context  [no-untyped-call]
 
 class A1:
     def __new__(cls): ...
@@ -267,3 +269,70 @@ class S:
     def f(self, f: int) -> str: ...
     @overload
     def f(self, f: str) -> int: ...
+
+
+[case testCheckUntypedDefs]
+# flags: --no-check-untyped-defs --allow-untyped-defs
+
+def foo(a=1):
+    1 + ""
+
+foo()  # E: Call to untyped function "foo" in typed context  [no-untyped-call]
+reveal_type(foo)  # E: Expression type contains "Any" (has type "(int) -> Untyped")  [no-any-expr] \
+                  # N: Revealed type is "def (a: int =) -> Untyped"
+
+def bar(a=1) -> None:
+    1 + ""  # E: Unsupported operand types for + ("int" and "str")  [operator]
+
+bar()
+reveal_type(bar)  # N: Revealed type is "def (a: int =)"
+
+
+[case inferPropertyTypes]
+# flags: --no-infer-function-types --no-default-return --no-check-untyped-defs
+class A:
+    @property
+    def f(self) -> int: ...
+    @f.setter
+    def f(self, value): ...
+    @f.deleter
+    def f(self): ...
+
+a = A()
+reveal_type(a.f)  # N: Revealed type is "int"
+a.f = ""  # E: Incompatible types in assignment (expression has type "str", variable has type "int")  [assignment]
+del a.f
+
+class B:
+    @property
+    def f(self): ...  # E: Function is missing a return type annotation  [no-untyped-def] \
+                      # N: Use "-> None" if function does not return a value \
+                      # E: Type of decorated function contains type "Any" ("(B) -> Untyped")  [no-any-decorated]
+    @f.setter
+    def f(self, value):
+        1 + ""
+    @f.deleter
+    def f(self): ...
+
+b = B()
+reveal_type(b.f)  # E: Expression has type "Untyped"  [no-any-expr] \
+                  # N: Revealed type is "Untyped"
+b.f = 1
+del b.f
+[builtins fixtures/property.pyi]
+
+
+[case testPropertyNoAnnotation]
+class A:
+    @property
+    def p(self): ...  # E: Property is missing a type annotation  [no-untyped-def]
+    @p.setter
+    def p(self, value): ...
+    @p.deleter
+    def p(self): ...
+
+a = A()
+reveal_type(a.p)  # N: Revealed type is "None"
+a.p = 1  # E: Incompatible types in assignment (expression has type "int", variable has type "None")  [assignment]
+del a.p
+[builtins fixtures/property.pyi]

--- a/test-data/unit/check-based-infer-function-types.test
+++ b/test-data/unit/check-based-infer-function-types.test
@@ -317,7 +317,7 @@ class B:
 b = B()
 reveal_type(b.f)  # E: Expression has type "Untyped"  [no-any-expr] \
                   # N: Revealed type is "Untyped"
-b.f = 1
+b.f = 1  # E: Usage of untyped name "f" in typed context  [no-untyped-usage]
 del b.f
 [builtins fixtures/property.pyi]
 
@@ -336,3 +336,34 @@ reveal_type(a.p)  # N: Revealed type is "None"
 a.p = 1  # E: Incompatible types in assignment (expression has type "int", variable has type "None")  [assignment]
 del a.p
 [builtins fixtures/property.pyi]
+
+
+[case testPropertyInheritInfer-xfail]
+# Overloaded overrides aren't supported yet
+class A:
+    @property
+    def p(self) -> int: ...
+class B(A):
+    @property
+    def p(self):
+        1 + ""  # E: Unsupported operand types for + ("int" and "str")  [operator]
+        return 1
+
+b = B()
+reveal_type(b.p)  # N: Revealed type is "int"
+c = b.p
+[builtins fixtures/property.pyi]
+
+
+[case testDefaultReturnOverride-xfail]
+class A:
+    def foo(self) -> int: ...
+    def bar(self, i: int) -> int: ...
+
+class B(A):
+    def foo(self): ...
+    def bar(self, i: int): ...
+
+b: B
+reveal_type(b.foo)  # N: Revealed type is "def () -> int"
+reveal_type(b.bar)  # N: Revealed type is "def (i: int) -> int"

